lib/sidekiq/delay_extensions/class_methods.rb

```diff
@@ -18,7 +18,11 @@ class DelayedClass
       def perform(yml)
         permitted_classes = Sidekiq::DelayExtensions.configuration.yaml_permitted_classes
         aliases = Sidekiq::DelayExtensions.configuration.yaml_aliases
-        (target, method_name, args) = YAML.load(yml, permitted_classes: permitted_classes, aliases: aliases)
+        if Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('3.1.0')
+          (target, method_name, args) = YAML.load(yml, permitted_classes: permitted_classes, aliases: aliases)
```

Comment:
since this could be a hotspot, better to calculate it once.  

Thanks for the idea here. If you're not sure how to safely define a Ruby version dependent yaml load, I'll take a peak at how we did it during our update

Reply:
I had done something like the below, though it couldn't be used as is. perhaps it should switch on `STDLIB_YAML.respond_to?(:unsafe_load)`



```ruby
require "yaml"

module Sidekiq
  module YAML
    STDLIB_YAML = ::YAML

    if !ENV["DEPENDENCIES_PREVIOUS"]
      # https://github.com/ruby/psych/blob/v4.0.3/lib/psych.rb#L271-L323
      # def self.unsafe_load(yaml, filename: nil, fallback: false, symbolize_names: false, freeze: false)
      def self.unsafe_load(yaml, **kwargs)
        STDLIB_YAML.unsafe_load(yaml, **kwargs)
      end

      # def self.safe_load(yaml, permitted_classes: [], permitted_symbols: [], aliases: false, filename: nil, fallback: nil, symbolize_names: false, freeze: false)
      def self.safe_load(yaml, **kwargs)
        STDLIB_YAML.safe_load(yaml, **kwargs)
      end

      # def self.unsafe_load_file(filename, **kwargs)
      def self.unsafe_load_file(filename, **kwargs)
        STDLIB_YAML.unsafe_load_file(filename, **kwargs)
      end

      # def self.safe_load_file(filename, **kwargs)
      def self.safe_load_file(filename, **kwargs)
        STDLIB_YAML.safe_load_file(filename, **kwargs)
      end
    else
      fail "Please use YAML.unsafe_load in #{__FILE__}" if STDLIB_YAML.respond_to?(:unsafe_load)
      # https://github.com/ruby/psych/blob/v3.1.0/lib/psych.rb#L271-L328
      # vs.
      # https://github.com/ruby/psych/blob/v4.0.3/lib/psych.rb#L271-L323
      # def self.load(       yaml, filename: nil, fallback: false, symbolize_names: false)
      #   vs.
      # def self.unsafe_load(yaml, filename: nil, fallback: false, symbolize_names: false, freeze: false)
      def self.unsafe_load(yaml, **kwargs)
        STDLIB_YAML.load(yaml, **kwargs)
      end

      # def self.safe_load(yaml, permitted_classes: [], permitted_symbols: [], aliases: false, filename: nil, fallback: nil, symbolize_names: false)
      #   vs.
      # def self.safe_load(yaml, permitted_classes: [], permitted_symbols: [], aliases: false, filename: nil, fallback: nil, symbolize_names: false, freeze: false)
      def self.safe_load(yaml, **kwargs)
        STDLIB_YAML.safe_load(yaml, **kwargs)
      end

      # def self.load_file filename, fallback: false
      #    vs.
      # def self.unsafe_load_file(filename, **kwargs)
      def self.unsafe_load_file(filename, **kwargs)
        STDLIB_YAML.load_file(filename, **kwargs)
      end

      # n/a
      #    vs.
      # def self.safe_load_file(filename, **kwargs)
      def self.safe_load_file(filename, **kwargs)
        STDLIB_YAML.load_file(filename, **kwargs)
      end
    end
  end
end
```

Reply:
I think that the right and easiest way is to check Ruby version.
`respond_to?(unsafe_load)` generally for me doesn't look like reliable solution, because we can't be sure what changes will be made in future, maybe this check become incorrect.
Also such approaches used in many gems and [Rails also](https://github.com/rails/rails/blob/90a6f3663b937cf90a7f642a10dfb36f8710ce6d/railties/lib/rails/ruby_version_check.rb#L5)

Reply:
That's different to check the ruby version vs gem version 

This is a psych issue, which is related to ruby version but not the same. The interface changes occasionally, but is pretty stable over the last decade, only adding ideas of safe and unsafe and changing a few method arguments.   I think duck typing is safer and appropriate here

Reply:
The version of the gem also [specified in Ruby sources](https://github.com/ruby/ruby/blob/feac15c1f3479f7d7641480a4608f3ac5d3b2fca/ext/psych/lib/psych/versions.rb#L5), but maybe you are right and it's better to check `Psych::VERSION`